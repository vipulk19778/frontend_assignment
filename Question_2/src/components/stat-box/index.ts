export { StatBox } from './stat-box';
