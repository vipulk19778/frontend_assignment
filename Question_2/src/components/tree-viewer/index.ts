export { TreeViewer } from './tree-viewer';
