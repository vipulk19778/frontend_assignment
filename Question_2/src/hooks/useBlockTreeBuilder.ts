import { useMemo } from 'react';

import { getPlantings, buildBlockTree } from '../utils/vineyard';

import type { BlockNode } from '../types/vineyard';

export const useBlockTreeBuilder = (): BlockNode[] => {
  const plantings = useMemo(() => getPlantings(), []);

  return useMemo(() => buildBlockTree(plantings), [plantings]);
};
