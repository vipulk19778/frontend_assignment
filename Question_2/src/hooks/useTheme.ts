export { useVineyardTheme as useTheme } from '../contexts/theme-context';
