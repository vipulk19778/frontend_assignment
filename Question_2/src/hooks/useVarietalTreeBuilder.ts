import { useMemo } from 'react';

import { getPlantings, buildVarietalTree } from '../utils/vineyard';

import type { VarietalNode } from '../types/vineyard';

export const useVarietalTreeBuilder = (): VarietalNode[] => {
  const plantings = useMemo(() => getPlantings(), []);

  return useMemo(() => buildVarietalTree(plantings), [plantings]);
};
