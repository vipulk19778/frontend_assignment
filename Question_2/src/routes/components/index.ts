export { ErrorBoundary } from './error-boundary';
export { RouterLink } from './router-link';
