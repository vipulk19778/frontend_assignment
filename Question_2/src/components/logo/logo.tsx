import type { LinkProps } from '@mui/material/Link';

import { useId } from 'react';
import { mergeClasses } from 'minimal-shared/utils';

import Link from '@mui/material/Link';
import { styled, useTheme } from '@mui/material/styles';

import { RouterLink } from 'src/routes/components';

import { logoClasses } from './classes';

// ----------------------------------------------------------------------

export type LogoProps = LinkProps & {
  isSingle?: boolean;
  disabled?: boolean;
};

export function Logo({
  sx,
  disabled,
  className,
  href = '/',
  isSingle = true,
  ...other
}: LogoProps) {
  const theme = useTheme();

  const gradientId = useId();

  const TEXT_PRIMARY = theme.vars.palette.text.primary;
  const PRIMARY_MAIN = theme.vars.palette.primary.main;
  const PRIMARY_DARKER = theme.vars.palette.primary.dark;

  const singleLogo = (
    <svg
      width="100%"
      height="100%"
      viewBox="0 0 80 80"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <linearGradient
          id={`${gradientId}-text`}
          x1="0%"
          y1="0%"
          x2="100%"
          y2="100%"
          gradientUnits="userSpaceOnUse"
        >
          <stop stopColor={PRIMARY_MAIN} />
          <stop offset="1" stopColor={PRIMARY_DARKER} />
        </linearGradient>
      </defs>
      <text
        x="40"
        y="50"
        textAnchor="middle"
        dominantBaseline="middle"
        fontSize="32"
        fontWeight="bold"
        fontFamily="Arial, sans-serif"
        fill={`url(#${gradientId}-text)`}
      >
        VE
      </text>
    </svg>
  );

  const fullLogo = (
    <svg
      width="100%"
      height="100%"
      viewBox="0 0 200 80"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <linearGradient
          id={`${gradientId}-text-full`}
          x1="0%"
          y1="0%"
          x2="100%"
          y2="100%"
          gradientUnits="userSpaceOnUse"
        >
          <stop stopColor={PRIMARY_MAIN} />
          <stop offset="1" stopColor={PRIMARY_DARKER} />
        </linearGradient>
      </defs>
      <text
        x="20"
        y="40"
        textAnchor="start"
        dominantBaseline="middle"
        fontSize="28"
        fontWeight="bold"
        fontFamily="Arial, sans-serif"
        fill={`url(#${gradientId}-text-full)`}
      >
        VE
      </text>
      <text
        x="60"
        y="40"
        textAnchor="start"
        dominantBaseline="middle"
        fontSize="16"
        fontWeight="500"
        fontFamily="Arial, sans-serif"
        fill={TEXT_PRIMARY}
      >
        Vineyard Explorer
      </text>
    </svg>
  );

  return (
    <LogoRoot
      component={RouterLink}
      href={href}
      aria-label="Logo"
      underline="none"
      className={mergeClasses([logoClasses.root, className])}
      sx={[
        {
          width: 40,
          height: 40,
          ...(!isSingle && { width: 200, height: 40 }),
          ...(disabled && { pointerEvents: 'none' }),
        },
        ...(Array.isArray(sx) ? sx : [sx]),
      ]}
      {...other}
    >
      {isSingle ? singleLogo : fullLogo}
    </LogoRoot>
  );
}

// ----------------------------------------------------------------------

const LogoRoot = styled(Link)(() => ({
  flexShrink: 0,
  color: 'transparent',
  display: 'inline-flex',
  verticalAlign: 'middle',
}));
