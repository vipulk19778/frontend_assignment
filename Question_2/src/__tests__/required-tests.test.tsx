import { renderHook, render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { ThemeProvider, createTheme } from '@mui/material/styles';

import { useBlockTreeBuilder } from '../hooks/useBlockTreeBuilder';
import { Logo } from '../components/logo';

// Create a theme with CSS variables support
const theme = createTheme({
  cssVariables: true,
  palette: {
    primary: {
      main: '#1976d2',
      dark: '#0d47a1',
    },
    text: {
      primary: '#000000',
    },
  },
});

const renderWithProviders = (ui: React.ReactElement) =>
  render(
    <BrowserRouter>
      <ThemeProvider theme={theme}>{ui}</ThemeProvider>
    </BrowserRouter>
  );

describe('Required Tests', () => {
  // Unit test for a tree hook
  it('should build block tree using hook', () => {
    const { result } = renderHook(() => useBlockTreeBuilder());

    expect(result.current).toBeInstanceOf(Array);
    expect(result.current.length).toBeGreaterThan(0);

    const firstBlock = result.current[0];
    expect(firstBlock).toHaveProperty('block');
    expect(firstBlock).toHaveProperty('varietals');
    expect(Array.isArray(firstBlock.varietals)).toBe(true);
  });

  // Render/snapshot test for a component
  it('should render Logo component snapshot', () => {
    const { container } = renderWithProviders(<Logo />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
