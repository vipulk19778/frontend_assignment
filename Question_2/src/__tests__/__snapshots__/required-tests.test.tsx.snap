// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Required Tests should render Logo component snapshot 1`] = `
<a
  aria-label="Logo"
  class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineNone ve__logo__root css-o10bd8-MuiTypography-root-MuiLink-root"
  data-discover="true"
  href="/"
>
  <svg
    fill="none"
    height="100%"
    viewBox="0 0 80 80"
    width="100%"
    xmlns="http://www.w3.org/2000/svg"
  >
    <defs>
      <lineargradient
        gradientUnits="userSpaceOnUse"
        id="«r0»-text"
        x1="0%"
        x2="100%"
        y1="0%"
        y2="100%"
      >
        <stop
          stop-color="var(--mui-palette-primary-main)"
        />
        <stop
          offset="1"
          stop-color="var(--mui-palette-primary-dark)"
        />
      </lineargradient>
    </defs>
    <text
      dominant-baseline="middle"
      fill="url(#«r0»-text)"
      font-family="Arial, sans-serif"
      font-size="32"
      font-weight="bold"
      text-anchor="middle"
      x="40"
      y="50"
    >
      VE
    </text>
  </svg>
</a>
`;
